File: src/main/store.ts
import { chmod, mkdir, readFile, rename, writeFile } from 'node:fs/promises'
import { dirname, join } from 'node:path'
import { defaultConfig, type CodexProxyConfig } from '../shared/types'
import { validateCodexAppPath, validateProxyConfig } from './core'

interface StoredSettings {
  config: CodexProxyConfig
  selectedAppPath?: string
}

export class SettingsStore {
  private readonly filePath: string

  constructor(userDataDirectory: string) {
    this.filePath = join(userDataDirectory, 'codex-proxy-settings.json')
  }

  async getConfig(): Promise<CodexProxyConfig> {
    const stored = await this.read()
    return stored?.config ?? { ...defaultConfig }
  }

  async saveConfig(value: unknown): Promise<CodexProxyConfig> {
    const config = validateProxyConfig(value)
    const current = await this.read()
    await this.write({
      config,
      ...(current?.selectedAppPath
        ? { selectedAppPath: current.selectedAppPath }
        : {}),
    })
    return config
  }

  async getSelectedAppPath(): Promise<string | undefined> {
    return (await this.read())?.selectedAppPath
  }

  async saveSelectedAppPath(value: string): Promise<void> {
    const selectedAppPath = validateCodexAppPath(value)
    const config = (await this.read())?.config ?? { ...defaultConfig }
    await this.write({ config, selectedAppPath })
  }

  private async read(): Promise<StoredSettings | undefined> {
    try {
      const text = await readFile(this.filePath, 'utf8')
      const value: unknown = JSON.parse(text)
      if (
        typeof value !== 'object' ||
        value === null ||
        Array.isArray(value)
      ) {
        return undefined
      }

      const record = value as Record<string, unknown>
      const config = validateProxyConfig(record.config)
      const selected =
        typeof record.selectedAppPath === 'string'
          ? validateCodexAppPath(record.selectedAppPath)
          : undefined
      return {
        config,
        ...(selected ? { selectedAppPath: selected } : {}),
      }
    } catch {
      return undefined
    }
  }

  private async write(settings: StoredSettings): Promise<void> {
    await mkdir(dirname(this.filePath), { recursive: true, mode: 0o700 })
    const temporaryPath = `${this.filePath}.tmp`
    await writeFile(
      temporaryPath,
      `${JSON.stringify(settings, null, 2)}\n`,
      {
        encoding: 'utf8',
        mode: 0o600,
      },
    )
    await chmod(temporaryPath, 0o600)
    await rename(temporaryPath, this.filePath)
  }
}

File: src-tauri/src/store.rs
use std::fs;
use std::path::{Path, PathBuf};

use serde::{Deserialize, Serialize};

use crate::core::{normalize_config, validate_app_path, validate_config};
use crate::types::CodexProxyConfig;

#[cfg(unix)]
use std::os::unix::fs::PermissionsExt;

#[derive(Clone)]
pub struct SettingsStore {
    file_path: PathBuf,
}

#[derive(Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
struct StoredSettings {
    config: CodexProxyConfig,
    #[serde(skip_serializing_if = "Option::is_none")]
    selected_app_path: Option<String>,
}

impl SettingsStore {
    pub fn new(app_data_dir: &Path) -> Self {
        Self {
            file_path: app_data_dir.join("codex-proxy-settings.json"),
        }
    }

    pub fn get_config(&self) -> CodexProxyConfig {
        self.read()
            .map(|settings| settings.config)
            .unwrap_or_default()
    }

    pub fn save_config(&self, mut config: CodexProxyConfig) -> Result<CodexProxyConfig, String> {
        normalize_config(&mut config)?;
        let selected_app_path = self.read().and_then(|settings| settings.selected_app_path);
        self.write(&StoredSettings {
            config: config.clone(),
            selected_app_path,
        })?;
        Ok(config)
    }

    pub fn selected_app_path(&self) -> Option<String> {
        self.read().and_then(|settings| settings.selected_app_path)
    }

    pub fn save_selected_app_path(&self, value: &Path) -> Result<(), String> {
        let selected = validate_app_path(value)?.to_string_lossy().into_owned();
        self.write(&StoredSettings {
            config: self.get_config(),
            selected_app_path: Some(selected),
        })
    }

    fn read(&self) -> Option<StoredSettings> {
        let text = fs::read_to_string(&self.file_path).ok()?;
        let mut settings: StoredSettings = serde_json::from_str(&text).ok()?;
        validate_config(&settings.config).ok()?;
        normalize_config(&mut settings.config).ok()?;
        if let Some(path) = settings.selected_app_path.as_deref() {
            validate_app_path(Path::new(path)).ok()?;
        }
        Some(settings)
    }

    fn write(&self, settings: &StoredSettings) -> Result<(), String> {
        let parent = self
            .file_path
            .parent()
            .ok_or_else(|| "配置目录无效。".to_string())?;
        fs::create_dir_all(parent).map_err(|error| error.to_string())?;
        #[cfg(unix)]
        fs::set_permissions(parent, fs::Permissions::from_mode(0o700))
            .map_err(|error| error.to_string())?;
        let temporary = self.file_path.with_extension("json.tmp");
        let text = format!(
            "{}\n",
            serde_json::to_string_pretty(settings).map_err(|error| error.to_string())?
        );
        fs::write(&temporary, text).map_err(|error| error.to_string())?;
        #[cfg(unix)]
        fs::set_permissions(&temporary, fs::Permissions::from_mode(0o600))
            .map_err(|error| error.to_string())?;
        fs::rename(temporary, &self.file_path).map_err(|error| error.to_string())
    }
}
